File: C++/reverseList.cpp
/**
 * @File    : reverseList.cpp
 * @Brief   : 反转链表
 * @Link    : https://leetcode-cn.com/problems/reverse-linked-list/
 * @Author  : Wei Li
 * @Date    : 2021-07-29
*/

#include <iostream>

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// 迭代方式
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        ListNode *prev = nullptr, *next;
        // 修改指针的指向即可
        while (head)
        {
            next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
};

// 递归方式
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        if (!head || !head->next)
        {
            return head;
        }
        ListNode *newHead = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return newHead;
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}

File: Python/reverseList.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: reverseList.py
@Function: 反转链表
@Link: https://leetcode-cn.com/problems/reverse-linked-list/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-29
"""


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# 迭代方式
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        pre, cur = None, head
        while cur:
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        return pre


# 递归方式
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        newHead = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return newHead


# ----------------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass
